File: KVLite/Configuration.cs
//
// Configuration.cs
// 
// Author(s):
//     Alessio Parma <alessio.parma@gmail.com>
//
// The MIT License (MIT)
// 
// Copyright (c) 2014-2015 Alessio Parma <alessio.parma@gmail.com>
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

using System;
using System.Configuration;

namespace PommaLabs.KVLite
{
    /// <summary>
    ///   TODO
    /// </summary>
    [Serializable]
    public sealed class Configuration : ConfigurationSection
    {
        private const string SectionName = "pommaLabs.kvlite";
        private const string DefaultStaticIntervalInDaysKey = "DefaultStaticIntervalInDays";
        private const string MaxCacheSizeInMBKey = "MaxCacheSizeInMB";
        private const string MaxLogSizeInMBKey = "MaxLogSizeInMB";
        private const string MaxCachedConnectionCountKey = "MaxCachedConnectionCount";
        private const string MaxCachedSerializerCountKey = "MaxCachedSerializerCount";
        private const string NancyCacheKindKey = "NancyCacheKind";
        private const string OperationCountBeforeSoftCleanupKey = "OperationCountBeforeSoftCleanup";

        private static readonly Configuration CachedInstance = (Configuration) ConfigurationManager.GetSection(SectionName);

        public static Configuration Instance
        {
            get { return CachedInstance; }
        }

        [ConfigurationProperty(DefaultStaticIntervalInDaysKey, IsRequired = false, DefaultValue = 30)]
        public int DefaultStaticIntervalInDays
        {
            get { return Convert.ToInt32(this[DefaultStaticIntervalInDaysKey]); }
        }

        [ConfigurationProperty(MaxCachedConnectionCountKey, IsRequired = false, DefaultValue = 10)]
        public int MaxCachedConnectionCount
        {
            get { return Convert.ToInt16(this[MaxCachedConnectionCountKey]); }
        }

        [ConfigurationProperty(MaxCachedSerializerCountKey, IsRequired = false, DefaultValue = 10)]
        public int MaxCachedSerializerCount
        {
            get { return Convert.ToInt16(this[MaxCachedSerializerCountKey]); }
        }

        [ConfigurationProperty(MaxCacheSizeInMBKey, IsRequired = true)]
        public int MaxCacheSizeInMB
        {
            get { return Convert.ToInt32(this[MaxCacheSizeInMBKey]); }
        }

        [ConfigurationProperty(MaxLogSizeInMBKey, IsRequired = true)]
        public int MaxLogSizeInMB
        {
            get { return Convert.ToInt32(this[MaxLogSizeInMBKey]); }
        }

        [ConfigurationProperty(NancyCacheKindKey, IsRequired = false)]
        public CacheKind NancyCacheKind
        {
            get { return (CacheKind) this[NancyCacheKindKey]; }
        }

        [ConfigurationProperty(OperationCountBeforeSoftCleanupKey, IsRequired = false, DefaultValue = 100)]
        public int OperationCountBeforeSoftCleanup
        {
            get { return Convert.ToInt16(this[OperationCountBeforeSoftCleanupKey]); }
        }
    }
}

File: Library/Configuration.vb
'
' Configuration.vb
' 
' Author(s):
'     Alessio Parma <alessio.parma@gmail.com>
'
' The MIT License (MIT)
' 
' Copyright (c) 2014-2015 Alessio Parma <alessio.parma@gmail.com>
' 
' Permission is hereby granted, free of charge, to any person obtaining a copy
' of this software and associated documentation files (the "Software"), to deal
' in the Software without restriction, including without limitation the rights
' to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
' copies of the Software, and to permit persons to whom the Software is
' furnished to do so, subject to the following conditions:
' 
' The above copyright notice and this permission notice shall be included in
' all copies or substantial portions of the Software.
' 
' THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
' IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
' FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
' AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
' LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
' OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
' THE SOFTWARE.

Imports System.Configuration

''' <summary>
'''   TODO
''' </summary>
Public NotInheritable Class Configuration
    Inherits ConfigurationSection
    
    Private Const SectionName As String = "KVLiteConfiguration"
    Private Const DefaultCachePathKey As String = "DefaultCachePath"
    Private Const MaxCacheSizeInMBKey As String = "MaxCacheSizeInMB"
    Private Const MaxCachedConnectionCountKey As String = "MaxCachedConnectionCount"
    Private Const OperationCountBeforeSoftCleanupKey As String = "OperationCountBeforeSoftCleanup"
    
    Private Shared ReadOnly CachedInstance As Configuration = CType(ConfigurationManager.GetSection(SectionName), Configuration)
    
    Public Shared ReadOnly Property Instance As Configuration
        Get
            Return CachedInstance
        End Get
    End Property

    <ConfigurationProperty(DefaultCachePathKey, IsRequired:=True)>
    Public ReadOnly Property DefaultCachePath As String
        Get
            Return CType(Me(DefaultCachePathKey), String)
        End Get
    End Property

    <ConfigurationProperty(MaxCachedConnectionCountKey, IsRequired:=False, DefaultValue:=10S)>
    Public ReadOnly Property MaxCachedConnectionCount As Short
        Get
            Return CType(Me(MaxCachedConnectionCountKey), Short)
        End Get
    End Property

    <ConfigurationProperty(MaxCacheSizeInMBKey, IsRequired:=True)>
    Public ReadOnly Property MaxCacheSizeInMB As Integer
        Get
            Return CType(Me(MaxCacheSizeInMBKey), Integer)
        End Get
    End Property

    <ConfigurationProperty(OperationCountBeforeSoftCleanupKey, IsRequired:=False, DefaultValue:=100S)>
    Public ReadOnly Property OperationCountBeforeSoftCleanup As Short
        Get
            Return CType(Me(OperationCountBeforeSoftCleanupKey), Short)
        End Get
    End Property
End Class
